File: Data Structures & Algorithms/daily-temperatures/submission-9.java
class Solution {
    public int[] dailyTemperatures(int[] temp) {
        Stack<Integer>st=new Stack<>();
        int n=temp.length;
        int[] ans=new int[n];
        Arrays.fill(ans,0);
        for(int i=n-1;i>=0;i--)
        {
            while(!st.empty() && temp[st.peek()]<=temp[i])
            {
                st.pop();
            }
            if(!st.empty())
            {
                ans[i]=st.peek()-i;
            }
            st.push(i);
        }
        return ans;
    }
}

File: Data Structures & Algorithms/products-of-array-discluding-self/submission-9.cpp
class Solution {
public:
    vector<int> productExceptSelf(vector<int>& nums) {
        int n=nums.size();
        vector<int>pre(n,1),su(n,1),ans(n);
        for(int i=1;i<n;i++)
        {
            pre[i]=pre[i-1]*nums[i-1];
        }
        for(int i=n-2;i>=0;i--)
        {
            su[i]=su[i+1]*nums[i+1];
        }
        for(int i=0;i<n;i++)
        {
            ans[i]=(su[i]*pre[i]);
        }
        return ans;
    }
};
